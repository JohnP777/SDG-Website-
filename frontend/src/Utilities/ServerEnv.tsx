export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}