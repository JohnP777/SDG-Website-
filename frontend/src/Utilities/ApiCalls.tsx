import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
