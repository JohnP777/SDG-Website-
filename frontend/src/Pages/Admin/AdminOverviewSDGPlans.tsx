import Page from '../../Components/Page';
import React, { useEffect, useState } from 'react';
import { Box } from '@mui/system';
import { TextField, Typography, IconButton } from '@mui/material';
import SDGPlansGraph from '../../Components/AdminSDGPlansGraph';
import AdminSDGActionPlan from '../../Components/AdminSDGActionPlan';
import SearchIcon from '@mui/icons-material/Search';
import FilterAltIcon from '@mui/icons-material/FilterAlt';
import FormControlLabel from '@mui/material/FormControlLabel';
import Checkbox from '@mui/material/Checkbox';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import { apiCallGet } from '../../Utilities/ApiCalls';
import { useNavigate } from 'react-router-dom'; 

interface ActionPlan {
  id: number;
  user_id: number;
  name_of_designers: string;
  impact_project_name: string;
  description: string;
  plan_content?: object;
  status?: string;
  created_at?: string;
  updated_at?: string;
}

const AdminOverviewSDGPlans = () => {
  const [actionPlans, setActionPlans] = useState<ActionPlan[]>([]);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [currentSearch, setCurrentSearch] = useState('');
  const open = Boolean(anchorEl);

  const navigate = useNavigate();

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  useEffect(() => {
    const fetchActionPlans = async () => {
      const response = await apiCallGet('api/admin/sdg-plans/', true);
      if (response?.statusCode === 200) {
        const plansArray = Object.values(response).filter(
          (item): item is ActionPlan =>
            typeof item === 'object' && item !== null && 'impact_project_name' in item
        );
        setActionPlans(plansArray);
      } else {
        console.error('Failed to fetch SDG action plans', response);
      }
    };
    fetchActionPlans();
  }, []);

  const filteredPlans = actionPlans.filter((plan) =>
    plan.impact_project_name.toLowerCase().includes(currentSearch.toLowerCase())
  );

  return (
    <Page>
      <Typography alignSelf='center' fontSize='2rem' fontWeight='500'>
        Admin Overview of SDG Action Plans
      </Typography>

      <SDGPlansGraph />

      <Typography variant='h5'>List of SDG Action Plans</Typography>

      <TextField
        placeholder='Search the forms'
        fullWidth
        onChange={(e) => setCurrentSearch(e.target.value)}
        sx={{
          backgroundColor: '#FFFFFF',
          padding: '5px',
          fontSize: '18px',
          borderRadius: '5px',
          boxSizing: 'border-box',
          '& fieldset': { border: 'none' },
          marginBottom: '20px',
          width: '50%'
        }}
        InputProps={{
          startAdornment: <SearchIcon sx={{ marginRight: '10px' }} />,
          endAdornment: (
            <IconButton onClick={handleClick}>
              <FilterAltIcon />
            </IconButton>
          )
        }}
      />

      <Menu
        id='basic-menu'
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{ 'aria-labelledby': 'basic-button' }}
      >
        <Typography pl={2}>Filter</Typography>
        <MenuItem>
          <FormControlLabel control={<Checkbox defaultChecked />} label='Teams' />
        </MenuItem>
        <MenuItem>
          <FormControlLabel control={<Checkbox defaultChecked />} label='Individual' />
        </MenuItem>
      </Menu>

      <Box sx={{ display: 'flex', gap: '10px', flexWrap: 'wrap' }}>
        {filteredPlans.map((plan, index) => (
          <AdminSDGActionPlan
            key={plan.id || index}
            actionPlanTitle={plan.impact_project_name}
            actionPlanOwner={plan.name_of_designers || 'Unknown'}
            actionPlanStatus={plan.status || 'Unknown'}
            onClick={() => navigate(`/sdg-form/${plan.id}`)} 
          />
        ))}
      </Box>
    </Page>
  );
};

export default AdminOverviewSDGPlans;