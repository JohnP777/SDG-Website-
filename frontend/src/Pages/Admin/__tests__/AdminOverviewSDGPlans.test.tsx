import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import AdminOverviewSDGPlans from '../AdminOverviewSDGPlans';
import { MemoryRouter } from 'react-router-dom';
import * as ApiCalls from "../../../Utilities/ApiCalls";

// Mock the API calls
jest.mock("../../../Utilities/ApiCalls", () => ({
  apiCallGet: jest.fn(),
  apiCallPatch: jest.fn(),
}));

// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};

describe('AdminOverviewSDGPlans', () => {
  beforeEach(() => {
    (ApiCalls.apiCallGet as jest.Mock).mockImplementation((url: string) => {
      if (url === 'api/admin/sdg-plans/') {
        return Promise.resolve({
          statusCode: 200,
          1: {
            id: 1,
            user_id: 101,
            name_of_designers: 'Alice',
            impact_project_name: 'Save the Whales',
            description: 'Protect marine life',
            status: 'In Progress',
          },
          2: {
            id: 2,
            user_id: 102,
            name_of_designers: 'Bob',
            impact_project_name: 'Plant More Trees',
            description: 'Reforesting efforts',
            status: 'Completed',
          },
        });
      }
    });
  });

  // Snapshot test
  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<AdminOverviewSDGPlans />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('fetches action plans', async () => {
    (ApiCalls.apiCallGet as jest.Mock).mockResolvedValue({
      statusCode: 200,
      1: {
        id: 1,
        user_id: 4,
        name_of_designers: 'Alice',
        impact_project_name: 'Save the Whales',
        description: 'Protect marine life',
        status: 'Draft',
      },
      2: {
        id: 2,
        user_id: 6,
        name_of_designers: 'Bob',
        impact_project_name: 'Plant More Trees',
        description: 'Reforesting efforts',
        status: 'Completed',
      },
    });

    renderWithRouter(<AdminOverviewSDGPlans/>)

    await waitFor(() => {
      expect(screen.getByText('Save the Whales')).toBeInTheDocument();
      expect(screen.getByText('Plant More Trees')).toBeInTheDocument();
    });

    const searchInput = screen.getByPlaceholderText('Search the forms');
    fireEvent.change(searchInput, { target: { value: 'Whales' } });

    await waitFor(() => {
      expect(screen.getByText('Save the Whales')).toBeInTheDocument();
      expect(screen.queryByText('Plant More Trees')).not.toBeInTheDocument();
    });
  });

});
