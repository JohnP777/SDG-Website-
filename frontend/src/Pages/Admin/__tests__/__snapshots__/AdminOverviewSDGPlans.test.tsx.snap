// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminOverviewSDGPlans should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-z3339j"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-u2vtxm-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: none;"
    >
      <div
        class="MuiContainer-root css-12pwiy3-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1t0x8et-MuiToolbar-root"
        >
          <a
            class="MuiTypography-root MuiTypography-subtitle1 css-1z0p208-MuiTypography-root"
            href="/"
          >
            SDG
            <br />
            ZOO
          </a>
          <div
            class="MuiBox-root css-dq25au"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-z77o6z-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xf0eoi-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiBox-root css-10s323d"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/"
              tabindex="0"
            >
              HOME
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/about-us"
              tabindex="0"
            >
              ABOUT
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-769nkz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              FEATURES
              <span
                class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="MuiBox-root css-1t0e2o"
          >
            <button
              aria-label="Profile settings"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1avsmgh-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bqh3fr-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-eyrs0v-MuiContainer-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-13j1ktd-MuiTypography-root"
      >
        Admin Overview of SDG Action Plans
      </p>
      <div
        class="MuiBox-root css-5k7ydn"
      >
        <div
          class="MuiFormControl-root css-qifc3l-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="filter-label"
          >
            Time Period
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-sc8y68-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="filter-label"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              role="combobox"
              tabindex="0"
            >
              Days
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
              tabindex="-1"
              value="Days"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w1u3ce"
              >
                <span>
                  Time Period
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="css-1gakkqb-MuiResponsiveChart-container"
        >
          <svg
            class="css-1evyvmv-MuiChartsSurface-root"
            height="400"
            viewBox="0 0 100 400"
            width="100"
          >
            <title />
            <desc />
            <defs />
            <g
              class="MuiChartsGrid-root css-g2sgfh-MuiChartsGrid-root"
            />
            <g
              clip-path="url(#:r7:-clip-path)"
            >
              <g />
              <g>
                <clippath
                  id=":ra:-auto-generated-id-0-line-clip"
                >
                  <rect
                    height="400"
                    width="50"
                    x="0"
                    y="0"
                  />
                </clippath>
                <g
                  clip-path="url(#:ra:-auto-generated-id-0-line-clip)"
                >
                  <path
                    class="MuiLineElement-root MuiLineElement-series-auto-generated-id-0 css-1srwpga-MuiLineElement-root"
                    cursor="unset"
                    d=""
                  />
                </g>
              </g>
              <text
                class="css-1f57y8b"
                x="50"
                y="200"
              >
                No data to display
              </text>
            </g>
            <g
              data-drawing-container="true"
            >
              <g />
            </g>
            <g
              class="MuiChartsLegend-root MuiChartsLegend-row css-nsbw82-MuiChartsLegend-root"
            >
              <g
                class="MuiChartsLegend-series MuiChartsLegend-series-auto-generated-id-0"
                transform="translate(55 25)"
              >
                <rect
                  class="MuiChartsLegend-itemBackground"
                  fill="transparent"
                  height="24"
                  style="pointer-events: none; cursor: unset;"
                  width="29"
                  x="-2"
                  y="-12"
                />
                <rect
                  class="MuiChartsLegend-mark"
                  fill="url(#myGradient)"
                  height="20"
                  style="pointer-events: none;"
                  width="20"
                  x="0"
                  y="-10"
                />
                <text
                  dominant-baseline="central"
                  style="pointer-events: none; font-family: \\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif; font-weight: 400; font-size: 1rem; line-height: 1; letter-spacing: 0.00938em; fill: rgba(0, 0, 0, 0.87);"
                  text-anchor="start"
                  x="25"
                  y="0"
                >
                  <tspan
                    dominant-baseline="central"
                    dy="0px"
                    x="25"
                  >
                    Total forms created
                  </tspan>
                </text>
              </g>
            </g>
            <clippath
              id=":r7:-clip-path"
            >
              <rect
                height="300"
                width="0"
                x="50"
                y="50"
              />
            </clippath>
            <defs>
              <lineargradient
                id="myGradient"
              >
                <stop
                  offset="5%"
                  stop-color="#a7c6fc"
                />
                <stop
                  offset="95%"
                  stop-color="#166afa"
                />
              </lineargradient>
            </defs>
          </svg>
        </div>
      </div>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
      >
        List of SDG Action Plans
      </h5>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cxicwd-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-zkc42u-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
            />
          </svg>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r8:"
            placeholder="Search the forms"
            type="text"
            value=""
          />
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="FilterAltIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M4.25 5.61C6.27 8.2 10 13 10 13v6c0 .55.45 1 1 1h2c.55 0 1-.45 1-1v-6s3.72-4.8 5.74-7.39c.51-.66.04-1.61-.79-1.61H5.04c-.83 0-1.3.95-.79 1.61"
              />
            </svg>
          </button>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiBox-root css-1f5vcwn"
      />
    </div>
  </div>
</DocumentFragment>
`;
